promtimer/cbstats.py

```diff
@@ -448,6 +450,83 @@ def get_user_log(cluster, user, password):
             return []
 
+class BackupStatsFiles(Source):
+    """
+    Represents a source of stats data that is a Prometheus instance running against
+    the stats data from cbbackupmgr runs.
+    """
+    def __init__(self, prometheus_tsdb_path, short_name, prometheus_port):
+        super().__init__(prometheus_port)
+        self._short_name = short_name
+        self._prometheus_tsdb_path = prometheus_tsdb_path
+        self._config = None
+
+    def short_name(self):
+        return self._short_name
+
+    def host(self):
+        return '127.0.0.1'
+
+    def maybe_start(self, log_dir):
+        """
+        Starts the Prometheus instance that serves stats for this source.
+        """
+        log_path = path.join(log_dir, f'prom-{self._short_name}.log')
+        listen_addr = '0.0.0.0:{}'.format(self.port())
+        args = [Source.PROMETHEUS_BIN,
+                '--config.file', path.join(util.get_root_dir(), 'noscrape.yml'),
+                '--storage.tsdb.path', self._prometheus_tsdb_path,
+                '--storage.tsdb.no-lockfile',
+                '--storage.tsdb.retention.time', '10y',
+                '--query.lookback-delta', '600s',
+                '--web.listen-address', listen_addr]
+        logging.info((f'starting prometheus server on {listen_addr} against '
+            f'{self._prometheus_tsdb_path}; logging to {log_path}'))
+        return util.start_process(args, log_path)
+
+    @staticmethod
+    def compute_min_and_max_times(archive_path):
+        """
+        Returns a 2-tuple containing an estimate of the min and max POSIX timestamps
+        times associated with this stats Source
+        :return: 2-tuple (min time, max time)
+        """
+        cpu_stats_dir = path.join(archive_path, 'logs', 'stats', 'cpu')
+
+        stat_files = []
+        for file in os.listdir(cpu_stats_dir):
+            if path.isfile(path.join(cpu_stats_dir, file)) and file[0] != '.':
+                stat_files.append(file)
+
+        timestamps = []
+        for file in stat_files:
+            # The timestamp in the filename indicates when the backup started
+            timestamps.append(datetime.datetime.fromtimestamp(int(file.split('-')[2])))
+        timestamps.sort()
+
+        if len(timestamps) == 0:
+            raise FileNotFoundError('No cpu stat files present in ' + cpu_stats_dir)
+
+        start_timestamp = timestamps[0] - datetime.timedelta(minutes=1) # This padding makes the graph look better
+
+        if len(timestamps) == 1:
+            last_filename_timestamp = timestamps[0]
+
+        if len(timestamps) > 1:
+            last_filename_timestamp = timestamps[-1]
+
+        last_stat_file = os.path.join(
+            cpu_stats_dir,
+            'repo-backup-' + str(int(time.mktime(last_filename_timestamp.timetuple())))
+        )
+
+        with open(last_stat_file, 'r') as f:
+            lines = f.read().splitlines()
+            last_timestamp = parse(lines[-1].split(';')[0])
```

Comment:
We should check that the file is in the right format before doing these indexes - i.e. check there is at least one line.